<a href="https://colab.research.google.com/github/Jorge2018/DataScience2022-2/blob/main/Script_matrizConfusion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [37]:
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns 
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingClassifier, RandomForestClassifier
from sklearn.metrics import confusion_matrix, classification_report

df=pd.read_csv(r'https://raw.githubusercontent.com/Jorge2018/RepositorioArchivos/main/cancer.csv')
df.head()



id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  
3          0.6638                  0.17300  
4          0.2364                  0.07678  

[5 rows x 32 columns]

**Encuentren los nombres de clase y determinen cómo están las clases balanceadas**

In [38]:
# al usar normalize = True con value_counts, nuestra salida es el porcentaje de cada clase (escrito como decimal)
df['diagnosis'].value_counts(normalize = True)

B    0.627417
M    0.372583
Name: diagnosis, dtype: float64

**Conviertan los nombres de cadenas de las clases para valores numéricos.**

In [39]:
df['diagnosis'] = df['diagnosis'].replace({'B': 0, 'M': 1})
df['diagnosis'] 

0      1
1      1
2      1
3      1
4      1
      ..
564    1
565    1
566    1
567    1
568    0
Name: diagnosis, Length: 569, dtype: int64

# Separa valores X, y

In [40]:
y = df['diagnosis']
X = df.drop(columns = 'diagnosis')
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

# Arbol de Decision

**Crea instancia del modelo y Revision de hiperparametros**

In [41]:
dec_tree = DecisionTreeClassifier(random_state = 42, max_depth=2)
dec_tree.get_params()

{'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': 2,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'random_state': 42,
 'splitter': 'best'}

**Entrenamiento, prediccion y evaluacion del modelo**

In [42]:
dec_tree.fit(X_train, y_train)
dec_tree.predict(X_test)
# calculen la exactitud de la clasificación
train_score = dec_tree.score(X_train, y_train)
test_score = dec_tree.score(X_test, y_test)
print(train_score)
print(test_score)



0.9460093896713615
0.916083916083916


# Bagging Classifier

**Crea instancia del modelo y Revision de hiperparametros**

In [43]:
bagreg = BaggingClassifier(random_state = 42)
bagreg.get_params()


{'base_estimator': None,
 'bootstrap': True,
 'bootstrap_features': False,
 'max_features': 1.0,
 'max_samples': 1.0,
 'n_estimators': 10,
 'n_jobs': None,
 'oob_score': False,
 'random_state': 42,
 'verbose': 0,
 'warm_start': False}

**Entrenamiento, prediccion y evaluacion del modelo**

In [44]:
bagreg.fit(X_train, y_train)
bagreg.predict(X_test)
# calculen la exactitud de la clasificación
train_score = bagreg.score(X_train, y_train)
test_score = bagreg.score(X_test, y_test)
print(train_score)
print(test_score)

0.9929577464788732
0.951048951048951


# RandomForestClassifier

**Crea instancia del modelo y Revision de hiperparametros**

In [45]:
rf = RandomForestClassifier(random_state = 42)
# Looking at some hyperparameters that seem tunable
rf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': 42,
 'verbose': 0,
 'warm_start': False}

**Entrenamiento, prediccion y evaluacion del modelo**

In [46]:
rf.fit(X_train, y_train)
rf.predict(X_test)
rf_train_score = rf.score(X_train, y_train)
rf_test_score = rf.score(X_test, y_test)
print(rf_train_score)
print(rf_test_score)

1.0
0.972027972027972


# Matriz de Confusion

In [47]:
pred=rf.predict(X_test)
matriz = confusion_matrix(pred,y_test)
print(matriz)

[[88  3]
 [ 1 51]]


In [48]:
# valores etiqueta cero
tp = matriz[0,0]
fp = matriz[1,0]
fn = matriz[0,1]
precision  = tp / (tp + fp)
recall     = tp / (tp + fn)
f1_  = 2*( precision * recall)/(precision + recall)

# valores etiqueta uno
tp = matriz[1,1]
fp = matriz[0,1]
fn = matriz[1,0]
precision1  = tp / (tp + fp)
recall1     = tp / (tp + fn)
f1_1  = 2*( precision * recall)/(precision + recall)


print(f'La precision para los valores con etiqueta 0 es:{precision}, el recall es:{recall}, y el f1 es:{f1_}')
print('\n')
print(f'La precision para los valores con etiqueta 1 es:{precision1}, el recall es:{recall1}, y el f1 es:{f1_1}')

La precision para los valores con etiqueta 0 es:0.9887640449438202, el recall es:0.967032967032967, y el f1 es:0.9777777777777779


La precision para los valores con etiqueta 1 es:0.9444444444444444, el recall es:0.9807692307692307, y el f1 es:0.9777777777777779


**validacion del cálculo, utilizando informe_clasificación de sklearn**

In [49]:
print('Comparativa de valores obtenidos para las etiquetas cero y uno:\n')
print(classification_report(pred,y_test))

Comparativa de valores obtenidos para las etiquetas cero y uno:

              precision    recall  f1-score   support

           0       0.99      0.97      0.98        91
           1       0.94      0.98      0.96        52

    accuracy                           0.97       143
   macro avg       0.97      0.97      0.97       143
weighted avg       0.97      0.97      0.97       143



**R1. La precision es del 94% aprox, para los datos de entrenamiento y del 92% aprox, para los datos de test, considerando los valores del arbol de decision**

**R2. para este ejercicio el mejor tipo de cancer es el benigno**

**R3. Los falsos positivos son 3**

**R4. Los falsos negativos son 1**

**R5. En lo particular creo que cualquier tipo de diagnostico de cancer es preocupante, si tuviera que elegir uno "peor" diria que es el falso negativo, ya que quiere decir que en realidad si tiene cancer**